### PR TITLE
Update trigger interfaces with callback, start, and stop notions.

### PR DESCRIPTION
**Pull request by jander-msft**
Refactor options providers into operations implementations.
Add proxy implementations for actions and triggers to allow the rest of the collection rule system to not have to understand the type of the options that are passed to actions and triggers.

### CODE REVIEW

File: src/Tools/dotnet-monitor/CollectionRules/Actions/ICollectionRuleActionProxy.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Actions
+{
+    /// <summary>
+    /// A collection rule action interface with untyped options that
+    /// provides is a proxy to the real action implementation and verifies
```
Comment (kkeirstead) [inline src/Tools/dotnet-monitor/CollectionRules/Actions/ICollectionRuleActionProxy.cs:13]: nit: typo ("provides is a proxy")

File: src/Tools/dotnet-monitor/CollectionRules/Triggers/ICollectionRuleTriggerFactoryProxy.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using System;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Triggers
+{
+    /// <summary>
+    /// A collection rule trigger factory interface with untyped options that
+    /// provides is a proxy to the real trigger factory implementation and verifies
```
Comment (kkeirstead) [inline src/Tools/dotnet-monitor/CollectionRules/Triggers/ICollectionRuleTriggerFactoryProxy.cs:12]: nit: same typo as above ("provides is a proxy")

File: src/Tools/dotnet-monitor/CollectionRules/Actions/ICollectionRuleActionOperations.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        /// </summary>
+        bool TryCreateOptions(
+            string actionName,
+            out object options);
+
+        /// <summary>
+        /// Attempts to validate an options instance of the options type
+        /// associated with the registered action name.
+        /// </summary>
+        void TryValidateOptions(
```
Comment (jander-msft) [inline src/Tools/dotnet-monitor/CollectionRules/Actions/ICollectionRuleActionOperations.cs:35]: Return `bool` or rename it `ValidateOptions`.

File: src/Tools/dotnet-monitor/CollectionRules/Triggers/ICollectionRuleTriggerOperations.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        /// </summary>
+        bool TryCreateOptions(
+            string triggerName,
+            out object options);
+
+        /// <summary>
+        /// Attempts to validate an options instance of the options type
+        /// associated with the registered trigger name.
+        /// </summary>
+        void TryValidateOptions(
```
Comment (jander-msft) [inline src/Tools/dotnet-monitor/CollectionRules/Triggers/ICollectionRuleTriggerOperations.cs:35]: Return `bool` or rename to `ValidateOptions`.

**Comment by kelltrick** [comment]
reviewed in meeting

**Review by kelltrick — APPROVED** [review]

**Pull request merged**